File: 02/main.js
const fs = require('fs');
const path = require('path');

const readInput = (file = '') => {
  const fileContent = fs.readFileSync(file, 'utf8');

  return fileContent;
};

const parseDirection = (line) => {
  const [dir, value] = line.split(' ');

  return [dir, Number(value)];
};

const partOne = (input = '') => {
  let hor = 0;
  let vertical = 0;

  input
    .split('\n')
    .map((line) => parseDirection(line))
    .forEach((v) => {
      const [dir, value] = v;

      if (dir === 'forward') {
        hor += value;
      }

      if (dir === 'down') {
        vertical += value;
      }

      if (dir === 'up') {
        vertical -= value;
      }
    });

  return hor * vertical;
};

const partTwo = (input = '') => {
  let hor = 0;
  let vertical = 0;
  let aim = 0;

  input
    .split('\n')
    .map((line) => parseDirection(line))
    .forEach((v) => {
      const [dir, value] = v;

      if (dir === 'forward') {
        hor += value;
        vertical += value * aim;
      }

      if (dir === 'down') {
        aim += value;
      }

      if (dir === 'up') {
        aim -= value;
      }
    });

  return hor * vertical;
};

const input = readInput(path.join(__dirname, 'input.txt'));
console.log(partOne(input));
console.log(partTwo(input));

File: 02/main.go
package main

import (
	"aoc-2021/util"
	"fmt"
	"path"
	"strconv"
	"strings"
)

func partOne(lines []string) int {
	hor := 0
	ver := 0

	for _, v := range lines {
		split := strings.Split(v, " ")
		dir := split[0]
		value, _ := strconv.Atoi(split[1])

		switch dir {
		case "forward":
			hor += value
		case "down":
			ver += value
		case "up":
			ver -= value
		}
	}

	return hor * ver
}

func partTwo(lines []string) int {
	hor := 0
	ver := 0
	aim := 0

	for _, v := range lines {
		split := strings.Split(v, " ")
		dir := split[0]
		value, _ := strconv.Atoi(split[1])

		switch dir {
		case "forward":
			hor += value
			ver += value * aim
		case "down":
			aim += value
		case "up":
			aim -= value
		}
	}

	return hor * ver
}

func main() {
	lines := util.ReadLines(path.Join("02", "input.txt"))

	fmt.Println(partOne(lines))
	fmt.Println(partTwo(lines))
}
